src/Ztm.WebApi/Watchers/TransactionConfirmation/EntityRuleRepository.cs

```diff
@@ -0,0 +1,247 @@
+        {
+            using (var db = this.db.CreateDbContext())
+            {
+                var rule = await db.TransactionConfirmationWatchingRules
+                    .Include(e => e.Callback)
+                    .Include(e => e.CurrentWatch)
+                    .FirstOrDefaultAsync(c => c.Id == id, cancellationToken);
+
+                return rule == null ? null : ToDomain(rule);
+            }
+        }
+
+        public async Task<TimeSpan> GetRemainingWaitingTimeAsync(Guid id, CancellationToken cancellationToken)
+        {
+            using (var db = this.db.CreateDbContext())
+            {
+                var rule = await db.TransactionConfirmationWatchingRules
+                    .Include(e => e.Callback)
+                    .FirstOrDefaultAsync(c => c.Id == id, cancellationToken);
+
+                if (rule == null)
+                {
+                    throw new KeyNotFoundException("The rule id is not found.");
+                }
+
+                return rule.RemainingWaitingTime;
+            }
+        }
+
+        public async Task<RuleStatus> GetStatusAsync(Guid id, CancellationToken cancellationToken)
+        {
+            using (var db = this.db.CreateDbContext())
+            {
+                var rule = await db.TransactionConfirmationWatchingRules
+                    .Include(e => e.Callback)
+                    .FirstOrDefaultAsync(c => c.Id == id, cancellationToken);
+
+                if (rule == null)
+                {
+                    throw new KeyNotFoundException("The rule id is not found.");
+                }
+
+                return (RuleStatus)rule.Status;
+            }
+        }
+
+        public async Task<IEnumerable<Rule>> ListWaitingAsync(CancellationToken cancellationToken)
+        {
+            using (var db = this.db.CreateDbContext())
+            {
+                var rules = await db.TransactionConfirmationWatchingRules
+                    .Include(e => e.Callback)
+                    .Include(e => e.CurrentWatch)
+                    .Where(e => e.Status == (int)WatchStatus.Pending && e.CurrentWatchId == null)
+                    .ToListAsync(cancellationToken);
+
+                return rules.Select(e => ToDomain(e));
+            }
+        }
+
+        public async Task SubtractRemainingWaitingTimeAsync(Guid id, TimeSpan consumedTime, CancellationToken cancellationToken)
+        {
+            using (var db = this.db.CreateDbContext())
+            {
+                var rule = await db.TransactionConfirmationWatchingRules
+                    .Include(e => e.Callback)
+                    .FirstOrDefaultAsync(c => c.Id == id, cancellationToken);
+
+                if (rule == null)
+                {
+                    throw new KeyNotFoundException("The rule id is not found.");
+                }
+
+                if (consumedTime < TimeSpan.Zero)
+                {
+                    throw new ArgumentException("The consumed time is negative.");
```

Comment:
Better to specify argument name too.

Reply:
Added and changed to support range exception